eole/modules/multi_headed_attn.py

```diff
@@ -460,12 +460,27 @@ def _compute_attention(
                         query,
                         key,
                         value,
-                        ~mask if mask is not None else None,
+                        ~mask if (mask is not None and not causal) else None,
```

Comment:
why are you adding "not causal" as a condition ?

Reply:
It should technically throw an error when is_causal and an attention mask is provided:
https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html

Base Python implementation does this, the CUDA backends don't.

Reply:
I know but I can't see how this can be right. At training time, on the decoder side, it is causal=True, still we have the mask.

Reply:
Stumbled upon this issue some time ago as well. I agree this is not clear.
Maybe we could make the condition on `causal` instead of `mask` to make things cleaner.
I don't have all the background of this module but I guess the point of supporting `is_causal` here is just in case we don't have a `mask`. (Not sure this case exists for now, but why not supporting it as long it's not hurting anything.)

Reply:
I had originally went with the the check on _causal_ as you propose @francoishernandez. I personally think it's the most clear.

Reply:
for sake of clarity let's make the code look like:
with sdpa_kernel([SDPBackend.MATH, SDPBackend.EFFICIENT_ATTENTION]):
                    attn_output = scaled_dot_product_attention(
                        query,
                        key,
                        value,
                        ~mask if mask is not None else None,
                        self.dropout_p,
                        is_causal=False,
                    )
I know is_causal is False by default in sdpa but at least it keeps things clearer.
The only case Model is "causal" is at training on decoder side for self-attn only (not context) and since our code passes a mask (sequential + triangular) it is ok to use the mask parameter in sdpa.